onnxruntime/featurizers_ops/cpu/truncated_svd_transformer.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/common.h"
+#include "core/framework/data_types.h"
+#include "core/framework/data_types_internal.h"
+#include "core/framework/op_kernel.h"
+
+#include "Featurizers/TruncatedSVDFeaturizer.h"
+#include "Featurizers/../Archive.h"
+
+namespace NS = Microsoft::Featurizer;
+
+namespace onnxruntime {
+namespace featurizers {
+
+template <typename T>
+struct TruncatedSVDTransformerImpl {
+  void operator()(OpKernelContext* ctx) const {
+    using MatrixT = NS::RowMajMatrix<T>;
+    using InputMatrixT = Eigen::Map<const MatrixT>;
+    // Create the transformer
+    Microsoft::Featurizer::Featurizers::TruncatedSVDTransformer<InputMatrixT> transformer(
+        [ctx]() {
+          const auto* state_tensor(ctx->Input<Tensor>(0));
+          const uint8_t* const state_data(state_tensor->Data<uint8_t>());
+
+          Microsoft::Featurizer::Archive archive(state_data, state_tensor->Shape().GetDims()[0]);
+          return Microsoft::Featurizer::Featurizers::TruncatedSVDTransformer<InputMatrixT>(archive);
+        }());
+
+    // Get the input
+    const auto* input_tensor(ctx->Input<Tensor>(1));
+    const T* input_data(input_tensor->template Data<T>());
+    // Matrix Eigen raw buffer mapping
+    const auto input_dim_0 = input_tensor->Shape()[0];
+    const auto input_dim_1 = input_tensor->Shape()[1];
+    InputMatrixT input_matrix(input_data, input_dim_0, input_dim_1);
+
+    // Prepare output shape which is [M, P] where P is the first dimension (rows)
+    // of P matrix from the transformer
+    const int64_t dim_0 = input_dim_0;
+    const int64_t dim_1 = transformer.getEigenVectorColsNumber();
+    TensorShape output_shape({dim_0, dim_1});
+    auto* output_tensor(ctx->Output(0, output_shape));
+    T* output_data = output_tensor->template MutableData<T>();
+    Eigen::Map<MatrixT> output_matrix(output_data, dim_0, dim_1);
+
+    std::function<void(MatrixT val)> callback;
+    callback = [&output_matrix](MatrixT val) {
```

Comment:
Is there any disadvantage in assigning this lambda in the previous line itself?

Reply:
The previous line would not be an assignment. It would require a constuctor which std::function does not have.

---
In reply to: [373741707](https://github.com/microsoft/onnxruntime/pull/2958#discussion_r373741707) [](ancestors = 373741707)

Reply:
Oh may be the constructor doesn't exist in the compiler version we're using? For reference https://en.cppreference.com/w/cpp/utility/functional/function/function

Reply:
Will play safe for older compilers.

---
In reply to: [373744508](https://github.com/microsoft/onnxruntime/pull/2958#discussion_r373744508) [](ancestors = 373744508,373741707)